### ISSUE TITLE
Allow matching of files to DATs via non-CRC32 checksums

### ISSUE THREAD

**Issue by TheBrainScrambler** [emmercm/igir#818]
### Describe the solution you'd like

From what I understand, igir checks if roms match a certain DAT by either comparing it to the CRC32 checksum stored in the header of a zip file, if the rom is zipped and the header exists, or by computing their CRC32 checksum.
I would like an option to force to verify roms by comparing them using the checksum algorithm of my choice, for me specifically SHA-256.

I was thinking of extending `igir report` by having a flag that allows me to specify that, maybe `--checksum-algo sha256`.

If there are no sha256 checksums in the DAT, then you try weaker checksums like SHA1, MD5 or CRC32, but you report that you couldn't verify the rom using the requested checksum algorithm. Or perhaps you just add a field saying with which checksum algorithm you did the verification. In any case, igir should first try the algorithm you asked for.

I'm wondering though: in https://igir.io/alternatives/ , you say that "ROMs: scan/checksum caching" is not supported by design. Does this concern my feature request ?
I was also thinking about the efficiency of this since it would require to unzip archives of roms to verify them. This is where I think having some kind of caching would be good, but then I know you don't want caching by design. Perhaps igir could make some kind of DAT file which for each rom could also contain a checksum of the zipped rom ? If you're using a deterministic zipping algorithm like TorrentZip then this would work.

**Comment by emmercm** [comment]
`igir` uses filesize in addition to CRC32, because of the implication here that CRC32 has more false positives than other hashing algorithms. Other ROM managers do this as well. This primarily helps when scanning a very large ROM collection as the chance of collision is not statistically insignificant with a large collection.

I think what you're really looking for here is the ability to use something other than CRC32+size when matching scanned files to a DAT, rather than during testing. For `igir test`, it would be extremely unlikely to receive a false positive CRC32+size.

For the file matching, it's going to be best to wait until I finally finish https://github.com/emmercm/igir/pull/740 which necessitates calculating the MD5 and SHA1 of scanned files.

---

RE: caching, you've touched on every important part that I've thought of. And if we consider CHDs to be a deterministic archive of sorts, they would apply here as well.

Thankfully `igir` can extract non-excessively large `.zip` files into memory without temp files (I just broke out https://github.com/emmercm/igir/issues/819 into its own issue), so calculating checksums on those isn't too bad.

Unfortunately, I don't see any active Node.js TorrentZip implementations, so the creation of them is unlikely to happen any time soon.

But in general, I think local caching can be handled separately from the issue of alternative checksum matching.

**Comment by TheBrainScrambler** [comment]
> I think what you're really looking for here is the ability to use something other than CRC32+size when matching scanned files to a DAT, rather than during testing. For `igir test`, it would be extremely unlikely to receive a false positive CRC32+size.

Yes, I maybe wasn't clear but this would be to match roms to DATs. So for `igir report` and not `igir test`

**Comment by emmercm** [comment]
File matching on something other than CRC32+size has been on my personal to-do list forever:

https://github.com/emmercm/igir/blob/7c6bfbb6993ddf83fd710c36c48a7a4cda92a087/src/modules/fileIndexer.ts#L40

My free time is going to be limited most likely through the end of the year, but I think this is a good suggestion. It should be easier to land after #819 has merged.

**Comment by emmercm** [comment]
Note: this will be required for processing MAME "disk" files, the MAME DATs only provide a SHA1 for those files and no filesize.

The code added in https://github.com/emmercm/igir/pull/835 will need to account for the zero filesize of these.

**Comment by emmercm** [comment]
Question for you @TheBrainScrambler while I wrap this feature up - what DAT group includes SHA256 in their DATs? It's very non-standard, and I haven't come across it yet myself.

**Comment by emmercm** [comment]
Interesting. I have absolutely no idea why their P/C DATs differ so greatly from their "standard" ones. I see no reason why P/C couldn't include SHA256 as well.

**Comment by TheBrainScrambler** [comment]
Thanks ! And any chance to see SHA256 support in igir as well ?

**Comment by emmercm** [comment]
@TheBrainScrambler absolutely, this will add it soon: https://github.com/emmercm/igir/pull/1032
